### PR TITLE
Switch to key code input on SDL: Part 1

### PR DESCRIPTION
**Pull request by Qrox**
#### Summary
SUMMARY: Interface "Bypass IME on SDL in most UIs"

#### Purpose of change
People sometimes use an IME to input non-English text, which manipulates raw keyboard input and converts it to unicode strings so people can input more characters than could be fitted onto a keyboard. The previous attempt at disabling IMEs in specific UIs (`ime_sentry`) was less than ideal because it's still possible to accidentally switch on IME mode and let it mess with your input. This is another attempt at it by switching to key code input on SDL, which IMEs (and possibly dead keys) do not interfere with.

#### Describe the solution
1. Added a new input event type `keyboard_code` (the original `keyboard` type was renamed to `keyboard_char`) to be used on non-android SDL. (I assume on android you get character/text input from software keyboards. Hardware keyboards might behave the same as on a PC but I don't have the environment to test, so I just disabled it entirely for android for now)
2. Added support for key modifiers, including serialization/deserialization. Also added code to migrate old keybinding config file.
3. Generate input events according to current UI's input mode (either old character/text input or new key code input), which is set in `input_context`.
4. Generate input events description according to current UI's input mode (this includes event description in keybindings menu, so only keybindings available in the current mode/platform are displayed).
5. Updated data/raw/keybindings.json to add new key code counterpart of the old keybindings.

It seems that curses doesn't have a portable way to receive raw key code input, so we cannot simply switch to key code mode on all platforms. And due to the fact that there are different keyboard layouts, we cannot simply convert SDL key code events to character input. So there has to be two modes. Fortunately with input context most of the UIs only handle actions rather than input events, which should minimize the drawback of using two distinct input modes.

#### Describe alternatives you've considered
Or I can just leave the system as is for consistency and let IME/dead key users suffer.

#### Testing
Tested most UIs using input_context and they were working as expected on SDL. Switched on a Chinese IME and there was no longer erratic input behavior in UIs that use the new key code mode. Also tested on curses and it worked as before.

#### Additional context
This is the first part. Some UIs that handle raw character/text input are marked to use the old input mode, to be migrated later to support both modes.

I'm still trying to find a way to also support text input in key code mode so we can use key code mode for all UIs on SDL, but text input events are fired independently of key code events, so it's probably hard to implement. Ideally all UIs should use the same mode for consistent keybindings, but if it turns out impossible, well, I guess it's better to have manageable inconsistency between a few UIs than to have IMEs to mess with your input throughout the entire game session.

### CODE REVIEW

**Comment by hexagonrecursion** [comment]
I sometimes accidentally enable caps lock and get confused by the flipped keybindings. It sounds like this will fix this issue by making the UI ignore the caps lock. Is this correct?

File: src/input.cpp [hunk 1]
```diff
@@ -180,6 +197,8 @@ void input_manager::init()
     }
 }
 
+static constexpr int current_keybinding_verion = 1;
```
Comment (BevapDin) [inline src/input.cpp:200]:
```suggestion
static constexpr int current_keybinding_version = 1;
```
Or what is the meaning of the word "verion"?

File: src/input.cpp [hunk 2]
```diff
@@ -222,26 +248,57 @@ void input_manager::load( const std::string &file_name, bool is_user_preferences
...
+            } else if( input_method == "keyboard_code" ) {
+                new_events[0].type = input_event_t::keyboard_code;
             } else if( input_method == "gamepad" ) {
-                new_event.type = input_event_t::gamepad;
+                new_events[0].type = input_event_t::gamepad;
             } else if( input_method == "mouse" ) {
-                new_event.type = input_event_t::mouse;
+                new_events[0].type = input_event_t::mouse;
+            } else {
+                keybinding.throw_error( "input_method", "unknown input_method" );
```
Comment (BevapDin) [inline src/input.cpp:265]:
```suggestion
                keybinding.throw_error( "unknown input_method", "input_method" );
```
The name of the problematic JSON object member is the second parameter.

File: src/input.cpp [hunk 3]
```diff
@@ -282,6 +350,7 @@ void input_manager::save()
                 jsout.start_object();
 
                 jsout.member( "id", action.first );
+                jsout.member( "ver", current_keybinding_verion );
```
Comment (BevapDin) [inline src/input.cpp:353]: Why not spell out the word "version"? What do we gain from this abbreviation?

**Comment by esotericist** [comment]
You sure don't stick to the small stuff, do you?

I think this is a solid step forward, and I'm glad to see you putting in the effort.

**Comment by Qrox** [comment]
@hexagonrecursion Yes, I just tested with caps lock on and input was not affected.

**Comment by ZhilkinSerg** [comment]
Works good to me. The only thing I am not fond of is textual representation of keybindings with modifiers, but that can be addressed later.

image

**Comment by Qrox** [comment]
Well, I think that's because the main UI hasn't be migrated to use the new system yet, so that's the old behavior in your screenshot.

**Pull request merged**